runtime/autoload/man.vim

```diff
@@ -11,127 +20,184 @@ catch /E145:/
+    call cursor(tag['lnum'], tag['col'])
   endif
+endfunction
 
-  silent exec editcmd.' man://'.page.(empty(sect)?'':'('.sect.')')
+" save current position in the stack
+function! s:push_tag() abort
+  let s:tag_stack += [{
+        \ 'buf':  bufnr('%'),
+        \ 'lnum': line('.'),
+        \ 'col':  col('.'),
+        \ }]
+endfunction
 
-  setlocal modifiable
-  silent keepjumps norm! 1G"_dG
-  if empty($MANWIDTH)
-    let $MANWIDTH = winwidth(0)
+" find the closest man window above/left
+" TODO(nhooyr) more ways to open manpages
+function! s:find_man(cmd) abort
+  if g:man_find_window == 0
+    return a:cmd
   endif
-  silent exec 'r!/usr/bin/man '.s:cmd(sect, page).' | col -b'
-  " Remove blank lines from top and bottom.
-  while getline(1) =~# '^\s*$'
-    silent keepjumps 1delete _
-  endwhile
-  while getline('$') =~# '^\s*$'
-    silent keepjumps $delete _
+  let thiswin = winnr()
+  while 1
+    if &filetype ==# 'man'
+      return 'edit'
+    endif
+    wincmd w
+    if thiswin == winnr()
+      return a:cmd
+    endif
   endwhile
-  setlocal nomodified
-  setlocal filetype=man
+endfunction
 
-  if invoked_from_man || editcmd ==# 'tabnew'
-    call s:set_window_local_options()
+" parses the page and sect out of 'page(sect)'
+function! s:parse_page_and_sect_fpage(fpage) abort
+  let ret = split(a:fpage, '(')
+  if len(ret) > 1
+    let iret = split(ret[1], ')')
+    return [ret[0], tolower(iret[0])]
   endif
+  return [ret[0], '']
 endfunction
 
-function s:set_window_local_options() abort
-  setlocal colorcolumn=0 foldcolumn=0 nonumber
-  setlocal nolist norelativenumber nofoldenable
+" parses the page and sect out of 'path/page.sect'
+function! s:parse_page_and_sect_path(path) abort
+  " regex for valid extensions that manpages can have
+  if a:path =~# '\.\%([glx]z\|bz2\|lzma\|Z\)$'
+    let tail = fnamemodify(fnamemodify(a:path, ':r'), ':t')
+  else
+    let tail = fnamemodify(a:path, ':t')
+  endif
+  let page = matchstr(tail, '^\f\+\ze\.')
+  let sect = matchstr(tail, '\.\zs[^.]\+$')
+  return [page, sect]
 endfunction
 
-function man#pop_page() abort
-  if s:man_tag_depth > 0
-    let s:man_tag_depth = s:man_tag_depth - 1
-    exec "let s:man_tag_buf=s:man_tag_buf_".s:man_tag_depth
-    exec "let s:man_tag_lin=s:man_tag_lin_".s:man_tag_depth
-    exec "let s:man_tag_col=s:man_tag_col_".s:man_tag_depth
-    exec s:man_tag_buf."b"
-    exec s:man_tag_lin
-    exec "norm! ".s:man_tag_col."|"
-    exec "unlet s:man_tag_buf_".s:man_tag_depth
-    exec "unlet s:man_tag_lin_".s:man_tag_depth
-    exec "unlet s:man_tag_col_".s:man_tag_depth
-    unlet s:man_tag_buf s:man_tag_lin s:man_tag_col
+function! s:read_page(sect, page, cmd)
+  execute s:find_man(a:cmd) 'man://'.a:page.(s:empty_sect(a:sect)?'':'('.a:sect.')')
+  setlocal modifiable
+  " TODO(nhooyr) perhaps do not load, merely redisplay?
+  " remove all the text, incase we already loaded the manpage before
+  keepjumps %delete _
+  let $MANWIDTH = winwidth(0)-1
```

Comment:
@justinmk do you know a better solution to this? 

http://stackoverflow.com/questions/26315925/get-usable-window-width-in-vim-script

I want 3 but I'm getting 1 here.